### PR TITLE
feat: add GitHub team import to new list page

### PR DESCRIPTION
**Pull request by brandonroberts**
## Description

- Adds GitHub Team import to new list creation page
- Adds additional org scope to auth flow

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents

Closes https://github.com/open-sauced/engineering/issues/179

## Mobile & Desktop Screenshots/Recordings

image

image

## Added tests?

- [ ] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [ ] 🙅 no documentation needed

### CODE REVIEW

File: pages/hub/lists/new.tsx [hunk 1]
```diff
@@ -197,6 +253,24 @@ const CreateListPage = () => {
             }}
           />
 
+          <InfoCard
+            title="Import your GitHub team"
```
Comment (bdougie) [inline pages/hub/lists/new.tsx:257]:
```suggestion
            title="Sync your GitHub team"
```

Small copy change to "sync." We should be able to update the team if someone leaves.

File: components/molecules/AuthSection/auth-section.tsx [hunk 2]
```diff
@@ -149,7 +149,10 @@ const AuthSection: React.FC = ({}) => {
             <Button
               variant="primary"
               onClick={async () =>
-                await signIn({ provider: "github", options: { redirectTo: `${host}${currentPath}` } })
+                await signIn({
+                  provider: "github",
+                  options: { redirectTo: `${host}${currentPath}`, scopes: "read:org" },
```
Comment (jpmcb) [inline components/molecules/AuthSection/auth-section.tsx:154]: Note: I had to sign out and sign back in on an incognito browser to get the higher scope auth flow to kick off.

**Review by jpmcb — APPROVED** [review]

**Review by nickytonline — CHANGES_REQUESTED** [review]

File: components/molecules/AuthSection/auth-section.tsx (hunk 2 above)
Comment (nickytonline) [reply components/molecules/AuthSection/auth-section.tsx:154]: I wonder if we need to indicate that they may need to log out, or I wonder if there's a way for us to get the higher auth scope to catch?

Comment (jpmcb) [reply components/molecules/AuthSection/auth-section.tsx:154]: It seems the whole auth flow needs to be re-triggered. I got this email once I did:

```
Hey jpmcb!

Previously, a third-party OAuth application (auth-beta-opensauced) had user:email scope. It was recently granted an additional scope (read:org). Visit ... for more information.

To see this and other security events for your account, visit https://github.com/settings/security-log

If you run into problems, please contact support by visiting https://github.com/contact

Thanks,
The GitHub Team
```

which makes me think that it'll fail until the higher auth is granted explicitly by the user.

Comment (brandonroberts) [reply components/molecules/AuthSection/auth-section.tsx:154]: Yes, that's true about the new auth flow. We could do an API check when they click the first button and redirect them if necessary, but thats adds a bit of friction.

File: lib/hooks/fetchGithubOrgTeams.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
+import { Octokit } from "octokit";
+import { supabase } from "lib/utils/supabase";
+
+interface GhOrgTeamResponse {
+  data: GhOrgTeam[];
+  status: number;
+  headers: {};
+}
+
+const fetchGithubOrgTeams = async (orgName: string | null) => {
```
[the remark this answers is not in the document]
Comment (brandonroberts) [reply lib/hooks/fetchGithubOrgTeams.ts:10]: These aren't hooks though as they are only fetching functions. If I rename them and try to call them, I get an error. I could move them into the `utils` folder.

File: components/organisms/GitHubTeamImportDialog/github-team-import-dialog.tsx [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+            </div>
+            <div className="flex pt-2">
+              <ToggleSwitch
+                name="followImported"
+                classNames="mr-2"
+                checked={follow}
+                handleToggle={() => setFollow((toFollow) => !toFollow)}
+              />
+              <Text className="leading-tight text-light-slate-9">Follow everyone imported.</Text>
+            </div>
```
Comment (nickytonline) [inline components/organisms/GitHubTeamImportDialog/github-team-import-dialog.tsx:109]: refactor (blocking): form elements need a label.

```suggestion
            <label className="flex mt-4 gap-4 items-center">
              <span>Organization</span>
              <SingleSelect
                options={userOrgs.map((userOrg) => ({
                  label: userOrg.organization_user.login,
                  value: `${userOrg.organization_user.login}`,
                }))}
                position="popper"
                value={selectedOrg}
                placeholder="Select an organization"
                onValueChange={(value) => setSelectedOrg(value)}
              />
            </label>
            <label className="flex gap-4 items-center">
              <span>Team</span>
              <SingleSelect
                options={teams.map((team) => ({ label: team.name, value: team.slug }))}
                position="popper"
                value={teamSlug}
                placeholder="Select a team"
                onValueChange={(value) => setTeamSlug(value)}
              />
            </label>
            <label className="flex">
              <ToggleSwitch
                name="followImported"
                classNames="mr-2"
                checked={follow}
                handleToggle={() => setFollow((toFollow) => !toFollow)}
              />
              <Text className="leading-tight text-light-slate-9">Follow everyone imported.</Text>
            </label>
```

**Before**

CleanShot 2023-10-25 at 17 17 08

**After**

CleanShot 2023-10-25 at 17 16 40

**Review by nickytonline — CHANGES_REQUESTED** [review]

File: components/organisms/GitHubTeamImportDialog/github-team-import-dialog.tsx (hunk 4 above)
Comment (nickytonline) [reply components/organisms/GitHubTeamImportDialog/github-team-import-dialog.tsx:109]: A bigger discussion @isabensusan, but we'll need to work on colour contrast issues with the text and backgrounds throughout the app.

File: lib/hooks/fetchGithubOrgTeams.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (nickytonline) [reply lib/hooks/fetchGithubOrgTeams.ts:10]: Ahh, I misunderstood. I got confused as they were in the hooks folder returning stuff. 🙃

**Review by nickytonline — APPROVED** [review]
Thanks for making the changes @brandonroberts! 🚀

File: components/molecules/AuthSection/auth-section.tsx (hunk 2 above)
Comment (bdougie) [reply components/molecules/AuthSection/auth-section.tsx:154]: This was enough friction from getting the Facebook OSPO from using the original open-sauced/open-sauced.

This is partly why installing the GitHub is the path we should introduce for progressively getting access to more read only data. 

I believe the on wording screen has a note about reading public data only. If not, it should.

**Review by OgDev-01 — APPROVED** [review]

**Review by bdougie — APPROVED** [review]
Email copy. Should be set up when we release this to production.

https://github.com/open-sauced/etl/issues/117

**Comment by brandonroberts** [comment]
I wonder if we can force a logout of all users also?

**Comment by nickytonline** [comment]
I think this would be the right way to go since it's a one click login since most people are logged in to GitHub already.

If you change the JWT secret, that should do it. All existing tokens will become invalid.

**Comment by jpmcb** [comment]
> I wonder if we can force a logout of all users also?

+1 - that would be ideal. Starting a thread with the engineers internally to get this coordinated.

**Pull request merged**